lib/types/authorization.d.ts

```diff
@@ -0,0 +1,11 @@
+// authorization.d.ts
+import type { IActionTypes } from '@types';
+import type { EAbilityStatus } from '@constants';
+
+export interface IAbility {
+  actions: IActionTypes[];
+  roles: EUserOrgRoles[];
+  contexts: IService<'name'>[];
+  name: string;
+  status: EAbilityStatus;
+}
```

Comment:
```
ICountryAbilities {
   [x in AbilityFuzziesT]: typeof value AbilityFuzziesT (generic)
}
```

AbilityFuzziesT = "age" | "idv" | "addv" | "role" | etc

countries: ['all', 'some', etc.]
age: ['all', "ranges", etc.]

Reply:
restrictions: ICountryAbilitiesMap

```
ICountryAbilitiesMap {
  [country]: ICountryAbilities
}
```

Reply:
`yourGeoRestrictions: ICountryAbilitiesMap`

```
ICountryAbilitiesMap {
  [country]: ICountryAbilities
}
```

Reply:
`yourOtherRestrictions: IAnyAbilityMap`

```
IAnyAbilityMap {
  [yourCriteria]: ICountryAbilities
}
```

Reply:
I can't think of a more optimised, yet cardinal data structure for this, thusof The Universal Fuzzy Ability algorithm.

Reply:
Then keyof user/org must include valueof IFuzzyAbilities for that key

Reply:
But that's not 2n, it's n + 1, because the array lives in the user/orgs. They usually have << length than criterias.

As in, user has 1 age.
User has 1 id.
User has 1 or few roles.
etc.

Reply:
The value of ability fuzzies type it generic because it can be strings `brazil`, but it can also be a date/point in time `new Date()`, or an object for aggregations! `date gte Today, country equals something`

Reply:
using this structure for both `restrictions` and `allowances` should account for the `OR` logic operand.

Reply:
e.g.
`users from brazil should be older than 18 years old`

`users from bahia can be older than 16`

Reply:
or, transposing:
`18+ years can be from brazil, italy, spain, germany, uk, etc`
`16+ years can be from etc`

Reply:
Map
`[age]: {
   [country]: [...countries]
   }`

Reply:
if restriction uses this map, then while added to any ability, will prevent users that match this criteria from seeing/taking the action.

Reply:
abilities can belong to projects, features or services

Reply:
`like-stuff`, `add-entry`, etc.

Reply:
abilities can map to roles, which is more performant than roles mapping to abilities.

Reply:
to check for roles when validating an action via an ability is n / 2 less complex than checking within all abilities when validating an action via a role.

Reply:
we have 6 roles, and possibly thousands of abilities.

Reply:
is the data redundant? you may ask. by a few bytes, yes.
nonetheless, the amount of abilities is << than the amount of the byte overhead might scale at.

for each new ability we're talking 1 (O) x few bytes (a few role names between 0 to 6 entries)
for each new user we'd be talking N(O) x considerably more bytes (one string for each ability) which can be flattened as N^2

Reply:
abilities tend to be << users.

Reply:
roles tend to be << abilities

Reply:
abilities per role tend to be >> user custom abilities

Reply:
so this can't be more optimised IMHO.